lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -36,30 +46,634 @@ class StructFieldCacheAnalysisPass : public ModulePass {
+  StructFieldAccessInfo* getStructFieldAccessInfo(const Type* T) const;
+  // Retrieve execution count for a basic block
+  Optional<uint64_t> getExecutionCount(const BasicBlock* BB) const{
+    Function* func = const_cast<Function*>(BB->getParent());
+    return LookupBFI(*func)->getBlockProfileCount(BB);
+  }
+  // Retrive a pair of information if the instruction is accessing any struct type and field number
+  Optional<std::pair<const Type*, unsigned> > getFieldAccessOnInstruction(const Instruction* I) const;
+  // Print all accesses of all struct types defined in the program
+  void debugPrintAllStructAccesses();
+  // Print the IR of the module with annotated information about struct access
+  void debugPrintAnnotatedModule();
+
+  // For stats
+  enum Stats{
+    StructPtrPtr,
+    FuncArgValue,
+    FuncArgNotDefined,
+    GepPassedIntoFunc,
+    GepPassedIntoBitcast,
+    GepUnknownUse,
+    UserNotInstructionNorOperator,
+    NoAccess,
+    PassedIntoOutsideFunction,
+    GepUsedOnStructPtr,
+    UnknownUsesOnStructPtr,
+    MaxNumStats
+  };
+
+  // Increment stats for one category
+  void addStats(unsigned Category) { StatCounts[Category]++; }
+  // Print a brief stats of struct access
+  void printStats();
+
+ private:
+  const Module& CurrentModule;
+  // Function reference that is used to retrive execution count for basic block
+  function_ref<BlockFrequencyInfo *(Function &)> LookupBFI;
+  std::unordered_map<const Type*, StructFieldAccessInfo*> StructFieldAccessInfoMap;
+  std::vector<unsigned> StatCounts;
+  const std::vector<std::string> StatNames = {
+    "Variable type is Struct**",
+    "Function argument is a value",
+    "Function argument is not defined in the program",
+    "GEP value passed into function calls",
+    "GEP value passed into bitcast",
+    "GEP value passed into unexpected opcode",
+    "User is not Instruction nor Operator",
+    "Struct defined but no accesses",
+    "Struct passed into functions defined out of scope",
+    "GEP instruction directly used on struct*",
+    "Unknown instruction directly used on struct*"
+  };
+  const std::vector<std::string> StructDefinitionTypeNames = {
+    "global struct",
+    "global struct*",
+    "global struct**",
+    "local struct",
+    "local struct*",
+    "local struct**"
+  };
+
+};
+
+class StructFieldAccessInfo
+{
+  /* This class is used to store all access information for each struct declared in
+    the program. It records all loads and stores to all fields of the struct to provide
+    essential information for cache-aware struct field analysis.
+  */
+ public:
+  StructFieldAccessInfo(const Type* T, const StructFieldAccessManager::StructDefinitionType ST, const Module& MD, const StructFieldAccessManager* M, const DICompositeType* D):
+      Eligiblity(true),
+      CurrentModule(MD),
+      StructureType(dyn_cast<StructType>(T)),
+      StructDefinition(ST),
+      DebugInfo(D),
+      NumElements(StructureType->getNumElements()),
```

Comment:
I checked type of T before calling constructor. I'll move the assignment of StructureType and NumElement into the constructor body to make sure they can be assigned.